src/radical/entk/task.py

```diff
@@ -29,1135 +114,227 @@ class Task(object):
+        'path'                 : '',
+        'tags'                 : None,
+        'rts_uid'              : None,
+        'parent_stage'         : {'uid': None, 'name': None},
+        'parent_pipeline'      : {'uid': None, 'name': None}
+    }
 
     # --------------------------------------------------------------------------
     #
     def __init__(self, from_dict=None):
 
-        self._uid     = ru.generate_id('task.%(counter)04d', ru.ID_CUSTOM)
-        self._name    = ""
-        self._state   = res.INITIAL
-        self._rts_uid = None
-
-        # Attributes necessary for execution
-        self._pre_exec   = list()
-        self._executable = ""
-        self._arguments  = list()
-        self._sandbox    = ""
-        self._post_exec  = list()
-        self._stage_on_error = False
-
-        self._lfs_per_process = 0
-        self._mem_per_process = 0
-        self._cpu_reqs        = {'processes'           : 1,
-                                 'process_type'        : None,
-                                 'threads_per_process' : 1,
-                                 'thread_type'         : None}
-        self._gpu_reqs        = {'processes'           : 0,
-                                 'process_type'        : None,
-                                 'threads_per_process' : 0,
-                                 'thread_type'         : None}
-
+        from_dict = from_dict or {}
+        if not isinstance(from_dict, dict):
+            raise ree.TypeError(expected_type=dict,
+                                actual_type=type(from_dict))
 
+        super().__init__()
 
-        # Data staging attributes
-        self._upload_input_data    = list()
-        self._copy_input_data      = list()
-        self._link_input_data      = list()
-        self._move_input_data      = list()
-        self._copy_output_data     = list()
-        self._link_output_data     = list()
-        self._move_output_data     = list()
-        self._download_output_data = list()
+        if not from_dict.get('uid'):
+            self['uid'] = ru.generate_id('task.%(counter)04d', ru.ID_CUSTOM)
+        self['state'] = res.INITIAL
 
-        # Name of file to write stdout and stderr of task
-        self._stdout = ""
-        self._stderr = ""
+        self.update(from_dict)
 
-        # Additional attributes that help in mapping tasks
-        # to cuds and cus to tasks
-        self._path      = None
-        self._exit_code = None
-        self._tags      = None
+    # --------------------------------------------------------------------------
+    #
+    def __eq__(self, other):
 
-        # Keep track of res attained
-        self._state_history = [res.INITIAL]
+        if not isinstance(other, Task):
+            return False
 
-        # Stage and pipeline this task belongs to
-        self._p_stage    = {'uid': None, 'name': None}
-        self._p_pipeline = {'uid': None, 'name': None}
+        return self['uid'] == other['uid']
```

Comment:
This seems like a confusing/dangerous shortcut. Wouldn't it be better to explicitly compare `task1.uid == task2.uid`? If not, could you be clear about the new comparison behavior in the class documentation?

Reply:
This is the same, since here we introduce Task as a Munch-based class (== extended typed dictionary), and `task1.uid` is the same as `task1['uid']`, and since it is inside the class definition accessing items is faster than accessing attributes

Reply:
I understand. I meant that, since this is a new comparison, it makes more sense to continue to compare the uid fields, explicitly, rather than to introduce a novel shorthand by which `task1 == task2` based only on its `uid` field. At the very least, such novel behavior comes with a burden of clear documentation.

Reply:
don't see any contradiction.. this method was added to be in sync with `__hash__`, and just to make it explicit that `uid` has to be unique. Tasks instances are considered equals if the have the same `uid`. You always can compare certain fields explicitly, e.g. `t1.uid == t2.uid`, `t1.name == t2.name`, `t1.as_dict() == t2.as_dict()`

Reply:
> Tasks instances are considered equals if the have the same uid.

Does this logic exist prior to this PR? I'm sorry if I overlooked that. I thought `Task.__eq__` and `Task.__hash__` were new with this change.

Reply:
> to make it explicit that uid has to be unique

It doesn't read that way to me. I read it more like "There may be multiple instances of Task that are assumed to refer to the same thing, but we have reasons to allow fields to be unpopulated or inconsistent, but we suppress the ability to check whether two instances are exact copies or not."

Reply:
If you want to clarify the uniqueness requirement, I think documentation should come before technical enforcement. Technical enforcement could reasonably be considered beyond the scope of this PR, if there wasn't previously something in place.

If you do want to enforce the uniqueness, it should occur through the creation methods or manager objects.

One option would be a class variable storing a WeakValueDictionary mapping UIDs to instances, and either refuse to allow copies to exist or come up with a scheme to manage them. But the precedent so far is that it is not the responsibility of Task to guarantee the uniqueness of the uids of its instances (right?), and I think that is fine.

Reply:
Original implementation does check uniqueness implicitly during the validation process ([here](https://github.com/radical-cybertools/radical.entk/blob/e2bc28a1ab5ac7197d9738b03bc3f3f8a0361ac1/src/radical/entk/task.py#L325))

I think what you describe is applicable if we would compare data structures, thus we say such objects are equal if all corresponding attributes are equal. In our case we check does two objects represent the same task, which is identified by ID (`uid`) and not by all set of attributes. EnTK (as well as RP) guarantees to create tasks with unique IDs within an application run.